### DISCUSSION TITLE
Order of operations for nested many-many updates

### DISCUSSION THREAD

**Discussion by SheaBelsky** [prisma/prisma#15867] [Q&A] [answered]
I am working with a schema that allows users to add tags to items. The schema looks something like this:

```
model User {
  // Meta
  id           String   @id @default(cuid())
  createdAt    DateTime @default(now())
  lastModified DateTime @updatedAt

  // Data
  tags AutocompleteOption[]
}

model AutocompleteOption {
  // Meta
  id           String   @id @default(cuid())
  createdAt    DateTime @default(now())
  lastModified DateTime @updatedAt

  // Data
  label String
  type AutocompleteType

  // The combination of label and type MUST be unique
  @@unique([label, type])
}

enum AutocompleteType {
  USER_TAG
}
```

When a user adds OR removes a tag from another user, I need to be able to disconnect all the old tags and then connect/create new tags based on the data of the new tags. To accomplish this, I am currently using `upsert` to EITHER connect the tag if it exists OR create the tag if it does not. (I am using `upsert` over `connectOrCreate` so I can update some information about the tag if it would be connected.) However, I don't know how to remove all the old tags before attempting to connect new ones. My original hypothesis was to try these things:

1. Pass `set: []` to the update operation
2. Pass `disconnect: [....]` to the update operation, then specify the IDs of all the old tags (which might be reconnected if they existed in the new set of tags)

However, it appears as though both `set` and `disconnect` occur AFTER the `upsert` operation occurs.

**My questions**
1. What's the order of operations for nested read/writes?
2. Is there a better way to **overwrite** or **replace** the value of `tags` on `Users`, rather than only being able to append to the end of an existing list?

**Comment by pantharshit00** [comment, marked as answer]
> 1. What's the order of operations for nested read/writes?

Order of operation is not guaranteed and it depends on the query graph that the prisma-engine will generate. I will recommend you to reset the state before trying to update the field in a separate query and then run both of your queries inside of a `$transaction([])`.


> 2. Is there a better way to overwrite or replace the value of tags on Users, rather than only being able to append to the end of an existing list?

I don't think this exists. You will need to use `set: []` or `deleteMany: {}` depends on if you want delete the related records, in a separate client call and then you will be able to add the values you want. I will recommend you to open a feature request about improving this here: https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
